File: Csharp/RT_CS/RT_CS/shapes/Shape.cs
﻿using RT_CS.Cameras;
using RT_CS.utility;
using RT_CS.Lights;
using RT_CS.Materials;
using System;
using System.Collections.Generic;
using System.Text;

namespace RT_CS.shapes
{
    public abstract class Shape
    {
        protected int sid;
        protected int materialId;

        public Shape(int id, int mid)
        {
            sid = id;
            materialId = mid;
        }

        public int GetMid()
        {
            return materialId;
        }

        public int GetId()
        {
            return sid;
        }
        private void IntersectShapes(Ray r, IntersectionRecord record, List<Shape> shapes)
        {
            foreach (Shape s in shapes)
            {
                s.Intersect(r, record);
            }
        }

       

        public abstract Vector3 GetNormal(Vector3 point);
        public abstract Vector2 GetUV(Vector3 pt);
        public abstract void Intersect(Ray r,IntersectionRecord intersection);
        public virtual Vector3 Paint(IntersectionRecord record, List<Light> lighting, Vector3 ambientLight, Vector3 viewDirection,
            int generation,List<Shape> shapes, List<Material> materials)
        {
             Shape GetShapeById(int id)
            {
                foreach (Shape shape in shapes)
                {
                    if (id == shape.GetId())
                        return shape;
                }
                return null;
            }

             Material GetMaterialById(int id)
            {
                foreach (Material material in materials)
                {
                    if (id == material.GetId())
                        return material;
                }
                return null;
            }
            Material mat = GetMaterialById(materialId);
            Vector2 uv = GetUV(record.Point);
            Vector3 ambientColor = mat.GetAmbientColor(uv);
            Vector3 diffuseColor = mat.GetDiffuseColor(uv);
            Vector3 specularColor = mat.GetSpecularColor(uv);
            float specularity = mat.GetSpecularity(uv);

            Vector3 color;
            color.x = 0;
            color.y = 0;
            color.z = 0;

            foreach (Light lit in lighting)
                color += LightupMaterial(lit, record, ambientColor, diffuseColor, specularColor, specularity, ambientLight, viewDirection,shapes, materials);

            color.x = MathF.Max(0,MathF.Min(1, color.x));
            color.y = MathF.Max(0,MathF.Min(1, color.y));
            color.z = MathF.Max(0,MathF.Min(1, color.z));

            Vector3 reflectivecolor, refractivecolor;
            reflectivecolor.x = 0;
            reflectivecolor.y = 0;
            reflectivecolor.z = 0;

            refractivecolor.x = 0;
            refractivecolor.y = 0;
            refractivecolor.z = 0;

            Vector3 reverseView = -1 * viewDirection;

            if (generation > 0)
            {
                float reflect = mat.GetReflectivity(uv);
                float transparent = mat.GetTransparency(uv);
                float refractiveIndex = mat.GetRefractiveIndex(uv);
                Vector3 normal = GetNormal(record.Point);

                if (reflect > 0)
                {
                    Vector3 reverseddir = (((2 * Vector3.Dot(reverseView, normal)) * normal) - (reverseView)).Normalize();
                    Ray r;
                    r.posit = record.Point;
                    r.direct = reverseddir;
                    IntersectionRecord reflectRec = new IntersectionRecord();
                    reflectRec.ExceptId = sid;
                    IntersectShapes(r, reflectRec, shapes);

                    if (reflectRec.Hit)
                    {
                        Shape reflectShape = GetShapeById(reflectRec.ShapeId);
                        reflectivecolor = reflectShape.Paint(reflectRec, lighting, ambientLight, reverseddir, generation - 1, shapes, materials);
                    }
                }

                if (transparent > 0)
                {
                    float cosIncidence = Vector3.Dot(normal, reverseView);
                    float sinIncidence = MathF.Sin(MathF.Acos(cosIncidence));
                    float sinRefractive = sinIncidence / refractiveIndex;
                    Vector3 refractedDirection = (-1 * ((reverseView/refractiveIndex) + (MathF.Cos(MathF.Asin(sinRefractive)) - (cosIncidence/refractiveIndex)) * normal)).Normalize();

                    Ray r;
                    r.posit = record.Point;
                    r.direct = refractedDirection;
                    IntersectionRecord refractRec = new IntersectionRecord();
                    refractRec.ExceptId = sid;
                    IntersectShapes(r, refractRec, shapes);

                    if (refractRec.Hit)
                    {
                        Shape refractShape = GetShapeById(refractRec.ShapeId);
                        refractivecolor = refractShape.Paint(refractRec, lighting, ambientLight, refractedDirection, generation - 1, shapes, materials);
                    }
                }
            }
            if (generation > 0)
                color = ((color * (1 - mat.GetReflectivity(uv) - mat.GetTransparency(uv))) + (mat.GetReflectivity(uv) * reflectivecolor) + (mat.GetTransparency(uv) * refractivecolor));

            return color;
        }

        private Vector3 LightupMaterial (Light lit, IntersectionRecord record, Vector3 ambientColor, Vector3 diffuseColor, Vector3 SpecularColor, float specularity, Vector3 ambientLight, Vector3 viewDirection, List<Shape> shapes, List<Material> materials)
        {
            Vector3 reverseLight = -1 * lit.GetDirection(record.Point);
            Ray shadowRay = new Ray() {  posit = record.Point, direct = reverseLight };
            IntersectionRecord shadowRec = new IntersectionRecord() { ExceptId = sid };
            IntersectShapes(shadowRay, shadowRec, shapes);
           
            Shape GetShapeById(int id)
            {
                foreach (Shape shape in shapes)
                {
                    if (id == shape.GetId())
                        return shape;
                }
                return null;
            }

            Material GetMaterialById(int id)
            {
                foreach (Material material in materials)
                {
                    if (id == material.GetId())
                        return material;
                }
                return null;
            }
            Vector3 norm = GetNormal(record.Point);

            Vector3 halfway = (viewDirection + reverseLight) * 0.5f;
            Vector3 lightColor = lit.GetColor(record.Point);
            float specularMult = MathF.Pow(Vector3.Dot(norm, halfway), specularity);
            float diffuseMult = MathF.Abs(Vector3.Dot(reverseLight, norm));
            float shadowMult;
            if (shadowRec.Hit)
            {
                float distSq = shadowRec.Distance * shadowRec.Distance;
          


                if (distSq < lit.LightPointDistance(record.Point))
                {
                    int newShapeId = shadowRec.ShapeId;
                    Shape newShape = GetShapeById(newShapeId);
                    Material newMaterial = GetMaterialById(newShape.GetMid());
                    Vector2 newUv = newShape.GetUV(shadowRec.Point);
                    float newTransparency = newMaterial.GetTransparency(newUv);
                    shadowMult = newTransparency;
                    diffuseMult *= shadowMult;
                    specularMult *= shadowMult;

                }
            }

            return Vector3.Mult(ambientLight,ambientColor) + diffuseMult * Vector3.Mult(lightColor ,diffuseColor) + specularMult * Vector3.Mult(lightColor, SpecularColor);
        }

    }

}

File: cpp/RT_CPP/RT_CPP/shape.cpp
#include "shape.h"
#include "operators.h"

using namespace shapes;

shape::shape(int id,int mid)
	: id(id), material_id(mid)
{

}

int shape::get_id() const
{
	return id;
}

int shape::get_material_id() const
{
	return material_id;
}

material* shape::get_material_by_id(int id, vector<material*>* materials)
{
	for (material* mat : *materials)
	{
		if (mat->get_id() == id)
		{
			return mat;
		}
	}
	return nullptr;
}

shape* shape::get_shape_by_id(int id, vector<shape*>* shapes)
{
	for (shape* shp : *shapes)
	{
		if (shp->get_id() == id)
		{
			return shp;
		}
	}
	return nullptr;
}

vector3 shape::paint(intersection_record& rec, vector<light*>& lights, vector3 ambientLight,vector3 viewDir,int generation, vector<shape*>* shapes, vector<material*>* materials)
{
	material* mat = get_material_by_id(material_id,materials);
	vector2 uv = get_uv(rec.point);
	vector3 ambientColor = mat->get_ambient_color(uv);
	vector3 diffuseColor = mat->get_diffuse_color(uv);
	vector3 specularColor = mat->get_specular_color(uv);
	float specularity = mat->get_specularity(uv);

	vector3 color = {0,0,0};
	for (light* l : lights)
	{
		color = color + lightup_material(l, rec, ambientColor, diffuseColor, specularColor, specularity, ambientLight, viewDir, shapes, materials);
	}
	color = color.clamp();
	vector3 reflectiveColor = {0,0,0};
	vector3 refractiveColor = {0,0,0};
	vector3 reversedView = -1 * viewDir;

	if (generation > 0)
	{
		float reflectivity = mat->get_reflectivity(uv);
		float transparency = mat->get_transparency(uv);
		float refractIndex = mat->get_refractive_index(uv);
		vector3 normal = get_normal(rec.point);
		if (reflectivity > 0)
		{
			vector3 reflectedDir = ( 2 * vector3::dot(normal, reversedView) * normal - reversedView).normalize();
			ray r;
			r.position = rec.point;
			r.direction = reflectedDir;
			intersection_record reflectRec;
			reflectRec.except_id = id;
			intersect_shapes(r, reflectRec, shapes);
			if (reflectRec.hit)
			{
				shape* reflectShape = get_shape_by_id(reflectRec.shape_id, shapes);
				reflectiveColor = reflectShape->paint(reflectRec, lights, ambientLight, reflectedDir, generation - 1, shapes, materials).clamp();
			}
		}
		if (transparency > 0)
		{
			float cosIncidence = vector3::dot(normal, reversedView);
			float sinIncidence = sinf(acosf(cosIncidence));
			float sinRefract = sinIncidence / refractIndex;
			float cosRefract = cosf(asinf(sinRefract));

			vector3 refractDir = ( - 1 * ((reversedView / refractIndex) + ((cosRefract - (cosIncidence / refractIndex)) * normal))).normalize();
			ray r;
			r.position = rec.point;
			r.direction = refractDir;
			intersection_record refractRec;
			refractRec.except_id = id;
			intersect_shapes(r, refractRec, shapes);
			if (refractRec.hit)
			{
				shape* refractShape = get_shape_by_id(refractRec.shape_id, shapes);
				refractiveColor = refractShape->paint(refractRec, lights, ambientLight, refractDir, generation - 1, shapes, materials).clamp();
			}
		}

		color = ((1 - transparency - reflectivity) * color) + (reflectivity * reflectiveColor) + (transparency * refractiveColor);
	}

	return color.clamp();
}


vector3 shape::lightup_material(light* light, intersection_record& rec,
	vector3 ambientColor, vector3 diffuseColor, vector3 specularColor, float specularity, vector3 ambientLight,vector3 viewDir, vector<shape*>* shapes,
	vector<material*>* materials)
{
	vector3 reversedLight = -1 * light->get_direction(rec.point);
	vector3 reversedView = -1 * viewDir;
	vector3 normal = get_normal(rec.point);
	vector3 lightColor = light->get_color(rec.point);
	ray rr;
	rr.position = rec.point;
	rr.direction = reversedLight;
	intersection_record recs;
	recs.except_id = id;
	intersect_shapes(rr, recs, shapes);
	float shadow_mult = 1;
	if (recs.hit)
	{
		float dist_square = recs.distance * recs.distance;
		if (dist_square < light->light_point_distance(rec.point))
		{
			vector2 uv2 = get_uv(recs.point);
			shape* shadow_shape = get_shape_by_id(recs.shape_id, shapes);
			material* shadow_mat = get_material_by_id(shadow_shape->get_material_id(), materials);
			float shadow_trans = shadow_mat->get_transparency(uv2);
			shadow_mult = shadow_trans;
		}
	}


	float diffuseMult = shadow_mult * fabsf(vector3::dot(reversedLight , normal));


	vector3 halfway = 0.5f * (reversedView + reversedLight);
	float specularMult =shadow_mult * powf( vector3::dot(halfway, normal), specularity);

	vector3 color = (vector3::mult(ambientColor,ambientLight)) + (  diffuseMult * vector3::mult(lightColor,diffuseColor)
		+ (specularMult * vector3::mult(lightColor,specularColor)));
	

	return color;

}

void shape::intersect_shapes(ray r, intersection_record& rec, vector<shape*>* shapes)
{
	for (shape* s : *shapes)
	{
		s->intersect(r, rec);
	}
}


